BLUESPAWN-client/src/util/permissions/permissions.cpp

```diff
@@ -291,6 +332,174 @@ namespace Permissions {
+			}
+		}
+		//Get privileges from groups that a user belongs to
+		if (otType == OwnerType::USER) {
+			PGROUP_USERS_INFO_0 pGroupInfo{ nullptr };
+			DWORD dEntriesRead{ 0 };
+			DWORD dEntriesTotal{ 0 };
+			NET_API_STATUS stat = NetUserGetLocalGroups(wDomainName.c_str(), wName.c_str(), 0, LG_INCLUDE_INDIRECT , reinterpret_cast<LPBYTE *>(&pGroupInfo), MAX_PREFERRED_LENGTH, &dEntriesRead, &dEntriesTotal);
+			AllocationWrapper awGroupInfoHandler{ pGroupInfo, 0, AllocationWrapper::NET_ALLOC };
+			if (stat != NERR_Success) {
+				LOG_ERROR("Error getting user groups. (Net Error: " << stat << ")");
+				return vPrivs;
+			}
+			//Add all privileges from groups to list of user's privileges
+			for (int i = 0; i < dEntriesRead; i++) {
+				Owner oGroup(pGroupInfo[i].grui0_name);
+				hr = LsaNtStatusToWinError(LsaEnumerateAccountRights(lPolicyHandle, oGroup.GetSID(), &pReceivedPrivs, &uPrivCount));
+				awReceivedPrivsHandler = { pReceivedPrivs, 0, AllocationWrapper::NET_ALLOC };
+				if (hr != ERROR_SUCCESS) {
+					LOG_ERROR("Error getting group privileges. (Error: " << GetLastError() << ")");
+				}
+				else if (hr == ERROR_SUCCESS) {
+					for (int i = 0; i < uPrivCount; i++) {
+						vPrivs.emplace_back(pReceivedPrivs[i]);
+					}
+				}
+			}
+		}
+		SetLastError(ERROR_SUCCESS);
+		return vPrivs;
+	}
+
+	bool Owner::HasPrivilege(IN const std::wstring& wPriv) {
+		auto OwnerPrivs = GetPrivileges();
+		return PrivListHasPrivilege(OwnerPrivs, wPriv);
+	}
+
+	bool Owner::PrivListHasPrivilege(IN const std::vector<LSA_UNICODE_STRING>& vPrivList, IN const std::wstring& wPriv) {
+		for (auto iter = vPrivList.begin(); iter != vPrivList.end(); iter++) {
+			if (wPriv.compare(iter->Buffer) == 0) return true;
+		}
+		return false;
+	}
+
+	std::vector<Owner> Owner::GetOwnersWithPrivilege(IN const std::wstring& wPriv) {
+		//Ensure policy handle is initialized
+		if (!bPolicyInitialized) {
+			InitializePolicy();
+			if (!bPolicyInitialized) {
+				LOG_ERROR("Error getting owners with privlige, couldn't initialize policy handle.");
+				return std::vector<Owner>{ };
+			}
+		}
+		LSA_UNICODE_STRING lPrivName = WStringToLsaUnicodeString(wPriv);
+		PLSA_ENUMERATION_INFORMATION pOwners{ nullptr };
+		ULONG uNumOwners{ 0 };
+		HRESULT hr = LsaNtStatusToWinError(LsaEnumerateAccountsWithUserRight(lPolicyHandle, &lPrivName, reinterpret_cast<PVOID *>(&pOwners), &uNumOwners));
+		AllocationWrapper awOwnersHandler{ pOwners, 0, AllocationWrapper::NET_ALLOC };
+		if (hr != ERROR_SUCCESS) {
+			LOG_ERROR("Error getting accounts with user privilege. (Error: " << hr << ")");
+			SetLastError(hr);
+			return std::vector<Owner>{ };
+		}
+		std::vector<Owner> vOwners;
+		for (int i = 0; i < uNumOwners; i++) {
+			DWORD dwSidLen = GetLengthSid(pOwners[i].Sid);
+			SecurityDescriptor sdSID = SecurityDescriptor::CreateUserSID(dwSidLen);
+			memcpy(sdSID.GetUserSID(), pOwners[i].Sid, dwSidLen);
+			Owner o(sdSID);
+			vOwners.emplace_back(o);
```

Comment:
weird C++ rule going on here - First, o is initialized with sdSID. Then, when you place it into vOwners, o is an l-value. This means that it can't be moved, resulting in the copy constructor being called. It's better to say vOwners.emplace_back({ sdSID });